airbyte-integrations/connectors/source-posthog/source_posthog/components.py

```diff
@@ -0,0 +1,55 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from dataclasses import dataclass
+from typing import Any, Mapping, MutableMapping, Optional
+
+from airbyte_cdk.sources.declarative.stream_slicers import SubstreamSlicer
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice, StreamState
+from airbyte_cdk.sources.declarative.requesters.request_options.interpolated_request_options_provider import (
+    InterpolatedRequestOptionsProvider,
+)
+
+
+@dataclass
+class PosthogIncrementalSlicer(SubstreamSlicer):
```

Comment:
we should be able to implement this using a cartesian stream slicer:
```
stream_slicer:
  type: Cartesian
  stream_slicers:
    - type: DatetimeStreamSlicer
       <...>
    - type: SubstreamSlicer
       parent_stream_configs:
          - stream: "*ref(definitions.projects_stream)"
              parent_key: "id"
              stream_slice_field: "project_id"
```

Reply:
I have completely re-written Events stream. 

Now it has the correct pagination and I added CartesianProductStreamSlicer, but it also required a lot of customization.

Reply:
Ok. I'm fine pushing this with custom components, but would you mind creating an issue describing what's missing from the native components so we know how to improve them for the next iteration?